### PR TITLE
[DataLoader] Inject the service container

### LINKED ISSUE #184: cannot use container in dataloader class
error is
`Fatal error: Call to a member function getParameter() on null in /var/www/myblog/src/AppBundle/DataFixtures/ORM/DataLoad.php on line 22`

fixtures function

```
public function getFixtures()
     {
         var_dump($this->container);
         $root_dir = $this->container->getParameter('kernel.root_dir');
         $upload_dir = $this->container->getParameter('upload_dir');
         $media_upload_path = $root_dir.'/../'.$upload_dir.'/media';
         $fs = new Filesystem();
         $finder = new Finder();
         var_dump($media_upload_path);
         if($fs->exists($media_upload_path)){
             $fs->remove($media_upload_path);
         }
         $fs->mkdir($media_upload_path,0777);
         return [];
     }
```
var_dump result is **NULL**

### PR DESCRIPTION
**Pull request by egeloen**
Hey!

This PR fixes #184 by injecting the container in the data loader.

### CODE REVIEW

File: tests/Doctrine/Finder/FixturesFinderTest.php [hunk 1]
```diff
@@ -27,6 +27,20 @@
...
+     */
+    private $finder;
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function setUp()
+    {
+        $this->finder = new FixturesFinder('DataFixtures/ORM');
+        $this->finder->setContainer($this->prophesize('Symfony\Component\DependencyInjection\ContainerInterface')->reveal());
```
Comment (theofidry) [inline tests/Doctrine/Finder/FixturesFinderTest.php]:
```php
$containerProphecy = $this->prophesize('Symfony\Component\DependencyInjection\ContainerInterface');
$containerProphecy->aMethod()->shouldNotBeCalled();
/* @var $container ContainerInterface */
$container = $containerProphecy->reveal();

$this->finder->setContainer($container);
```

This changes two things:

- Ensure no call is being made on the container
- Don't raise a scrutinizer type error

File: src/Doctrine/Finder/FixturesFinder.php [hunk 2]
```diff
@@ -21,9 +23,22 @@
...
 {
     /**
+     * @var ContainerInterface
+     */
+    private $container;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setContainer(ContainerInterface $container = null)
```
Comment (theofidry) [inline src/Doctrine/Finder/FixturesFinder.php]: why making it nullable?

File: src/Doctrine/Finder/FixturesFinder.php [hunk 3]
```diff
@@ -21,9 +23,22 @@
  *
  * @author Théo FIDRY <theo.fidry@gmail.com>
  */
-class FixturesFinder extends \Hautelook\AliceBundle\Finder\FixturesFinder
+class FixturesFinder extends \Hautelook\AliceBundle\Finder\FixturesFinder implements ContainerAwareInterface
 {
     /**
+     * @var ContainerInterface
```
Comment (theofidry) [inline src/Doctrine/Finder/FixturesFinder.php]:
```php
@var ContainerInterface|null
```

File: src/Doctrine/Finder/FixturesFinder.php [hunk 4]
```diff
@@ -96,7 +111,11 @@ private function getDataLoadersFromDirectory($path)
 
             if (true === isset($phpClasses[$sourceFile])) {
                 if ($reflectionClass->implementsInterface('Hautelook\AliceBundle\Doctrine\DataFixtures\LoaderInterface')) {
-                    $loaders[$className] = new $className();
+                    $loaders[$className] = $loader = new $className();
```
Comment (theofidry) [inline src/Doctrine/Finder/FixturesFinder.php]: not a bit fan of that, it's neater and more concise but harder to read as well:

```php
$loader = new $className();
$loaders[$className] = $loader;
```

File: src/Doctrine/Finder/FixturesFinder.php [hunk 5]
```diff
@@ -96,7 +111,11 @@ private function getDataLoadersFromDirectory($path)
...
+
+                    if ($reflectionClass->implementsInterface('Symfony\Component\DependencyInjection\ContainerAwareInterface')) {
```
Comment (theofidry) [inline src/Doctrine/Finder/FixturesFinder.php]: instead you could use

```php
if ($loader instanceof ContainerAwareInterface) {
  // ...
}
```

**Comment by theofidry** [comment]
Thanks for it @egeloen, some small changes and this can be merged :)

**Comment by egeloen** [comment]
@theofidry All comments have been applied :)

File: tests/Doctrine/Finder/FixturesFinderTest.php [hunk 6]
```diff
@@ -27,6 +28,27 @@
...
+    private $finder;
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function setUp()
+    {
+        $this->finder = new FixturesFinder('DataFixtures/ORM');
+
+        $container = $this->prophesize('Symfony\Component\DependencyInjection\ContainerInterface');
```
Comment (theofidry) [inline tests/Doctrine/Finder/FixturesFinderTest.php]: `$containerProphecy`, it's a prophecy object not a stub. It's better to keep both as you can still do calls on the prophecy object even after you got your stub, ex:

```php
$containerProphecy = $this->prophesize('Symfony\Component\DependencyInjection\ContainerInterface');
$container->get()->willReturn('yo');

/* @var $container ContainerInterface */
$container = $container->reveal();

$this->finder->setContainer($container);

$containerProphecy->get()->shouldHaveBeenCalledTimes(1);
```

Also it doesn't matter at which point you create the stub:

```php
$containerProphecy = $this->prophesize('Symfony\Component\DependencyInjection\ContainerInterface');
$container->get()->willReturn($containerProphecy->reveal()); // if get() was fluent for example (which is non sense in this case, it's just for the example)

/* @var $container ContainerInterface */
$container = $container->reveal();

$this->finder->setContainer($container);

$containerProphecy->get()->shouldHaveBeenCalledTimes(1);
```

Comment (egeloen) [inline tests/Doctrine/Finder/FixturesFinderTest.php]: I'm sorry but I don't get your point here, I usually rely on PHPUnit mock... How would you like I update this snippet? :)

Comment (theofidry) [inline tests/Doctrine/Finder/FixturesFinderTest.php]: Renaming `$container = $this->prophesize('Symfony\Component\DependencyInjection\ContainerInterface')`  to `$containerProphecy` instead and keep `$container` for `$containerProphecy->reveal()`, i.e. not mixing the two different objects.

TL;DR

My point is that the object return by `$this->prophesize('Symfony\Component\DependencyInjection\ContainerInterface')` is a prophecy object, whereas `$this->prophesize('Symfony\Component\DependencyInjection\ContainerInterface')->reveal()` is a stub.

For example in PHPUnit Mock:

```php
$stub = $this->getMockBuilder('SomeClass')
                     ->getMock();
```

so you have a mock builder and the stub. With Prophecy it's kind of similar. There is some subtle differences between PHPUnit Mock and Prophecy, but it's better explained [here](http://everzet.com/post/72910908762/conceptual-difference-between-mockery-and-prophecy) ;)

**Comment by egeloen** [comment]
@theofidry PR ready to merge :)

**Comment by theofidry** [comment]
Nice! Thanks very much :)
